### PR TITLE
Article 112: compiler

### PR DESCRIPTION
**Pull request by tschudin**
#### Length of text ####

:book:  long

####  Difficulty to read ####
:mouse:  light

####  Difficulty to understand #### 
:school_satchel:  easy

####  Type of text #### 

:bowtie:  informative

####  Type of header #### 
 
:bulb:  informative

### CODE REVIEW

**Comment by Teifun2** [comment]
**Plenum:**
Problems:
- What is a compiler, should be explained.
- There is a similiar article (combine?)


Ideas:
- How to do the image? (text references a sepcific image)
- PR should mention the (picture is on the left) / cut (is on the left)

**Comment by tschudin** [comment]
Der importierte Artikel hat die Kommentare schon aufgenommen, wie folgt:
- beide gestellten Fragen wurden in einem Zusatzsatz erledigt: was ein Compiler macht (Uebersetzen), und Inhalt des ähnlichen Artikels (Maschinen-Code).
- zusätzlich wurde auch JavaScript-Transpilation als Beispiel eingeführt: das Bild passt auch auf diesen Fall.
- Der Verweis auf den Ort des Bildes konnte problemlos entfernt werden.

**Comment by Teifun2** [comment]
- "The picture is hers and shows how a computer program can produce another computer program." 
This sentence suggests that the "compiler" can itself produce another computer program which is not directly the case. Rather a compiler translates from one language to another.
- The different programming languages mentioned at the end could be confusing.

**Comment by tschudin** [comment]
* replaced "create" by "translate a program"
* I left the JS6 to JS5 case, but made an independent sentence and singled out the word transpiler. This justifies mentioning the "funny" list of compiler variants.

**Comment by Teifun2** [comment]
We agree with the changes made to the article, but we think the article lacks the information of the most important and basic aspect of a compiler, transforming human readable code into machine language. 
If you would like we would consider a discussion in the plenum.

**Comment by tschudin** [comment]
@Teifun2 :
- I added "human-written" to address the concern.

It's not about human-_readable_ at all: Just to counter the claim of what a compiler's "most important and basic aspect" is: the input to a compiler does not have to be human-readable. A hard core example is the [C obfuscation contest](https://ioccc.org/), another one being assembler inline code to C programs, as used in the Linux kernel.

I don't think this needs a plenum discussion but you can bring it up there, of course.

**Comment by tschudin** [comment]
@MoiraZuber : I don't think this needs a new picture (the hidden agenda of this article is to show the historic dimension), just a reference (link), the artist's name and the year. What do you think?

**Comment by Teifun2** [comment]
**Plenum:**
- Drop last sentence? Drop example!
- Dont focus on C code (Human-written C)
- Change Title

**Comment by tschudin** [comment]
addressed all three remarks - back to review state.

**Comment by MoiraZuber** [comment]
I am not sure what to do with the image. Can it legally be used like this? Or should I make an abstraction?

**Comment by marcelluethi** [comment]
@tschudin We would like to promote the article to the website, but there still is a question regarding the image: We would prefer to use the original image, but are unsure about the copyright situation. How should we proceed?

File: articles/article-112-compiler.md [hunk 1]
```diff
@@ -0,0 +1,32 @@
+<!-- BEGIN TITLE -->
+# We have Machines within Machines
+<!-- END TITLE -->
+
+<!-- BEGIN BODY -->
+In 1954, computers just computed numbers. Grace Hopper, a famous
```
Comment (noverney) [inline articles/article-112-compiler.md]: Nothing grammatically incorrect but to be more concise you could say: 

Grace Hopper, then a famous female programmer, had a hard time convincing her colleagues that computers can do more. 
- I cut away those machines as computers can just be repeated

File: articles/article-112-compiler.md [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+female programmer around then, had a hard time to convince her
+colleagues that these machines can do more.
+
+The picture is hers and shows how a computer program can translate
```
Comment (noverney) [inline articles/article-112-compiler.md]: She made the picture below demonstrating how a computer program can translate from one computer language to another. 
- translate from .. to .. 
- I am assuming you mean translating from languages not programs

File: articles/article-112-compiler.md [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+one computer program into another one. She even coined the term that is still in use
+today: a **compiler**.
+
+Nowadays, we have many more of these translation and transformation
+machines, with names as funny as the picture: _cross-compilers_,
+_transpilers_, and _compiler-compilers_ (cocos). Their mission is to
+translate from one computer language to another, for example
```
Comment (noverney) [inline articles/article-112-compiler.md]: I realized I made a correction that has the exact same line as what you have hear, in which case you can just drop that part and say:
These compilers can translate many things, for example...

**Comment by tschudin** [comment]
picked some suggested edits but not all.

- kept "machines" in 1st paragraph, (otherwise the title is not understood)
- "translate prog to prog" in 2nd, (progs translating progs to other progs is the take-home msg)
- "language to another" in 3rd paragraph, for generalization

**Pull request closed without merging**